hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationCommandCreator.java

```diff
@@ -0,0 +1,239 @@
+      final PlacementPolicy containerPlacement, final ConfigurationSource conf,
+      NodeManager nodeManager) {
+    this.containerPlacement = containerPlacement;
+    this.currentContainerSize = (long) conf
+        .getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+            ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.nodeManager = nodeManager;
+  }
+
+  /**
+   * Identify a new set of datanode(s) to reconstruct the container and form the
+   * SCM command to send it to DN. In the case of decommission, it will just
+   * generate the replicate commands instead of reconstruction commands.
+   *
+   * @param replicas - An instance of ContainerReplicaCount, containing the
+   *                   current replica count and inflight adds and deletes
+   * @param pendingOps - Inflight replications and deletion metrics.
+   * @param result - Health check result.
+   * @param remainingMaintenanceRedundancy - represents that how many nodes go
+   *                                      into maintenance.
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final ContainerReplicaPendingOps pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas,
+            pendingOps.getPendingOps(container.containerID()),
+            remainingMaintenanceRedundancy);
+    ContainerHealthResult containerHealthResult =
+        new ECContainerHealthCheck().checkHealth(container, replicaCount);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+
+    if (containerHealthResult.getHealthState()
+        != ContainerHealthResult.HealthState.UNDER_REPLICATED) {
+      LOG.info(
+          "The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      // Assert replicaCount also telling it's sufficiently replicated.
+      return null;
+    }
+    final ContainerID id = container.containerID();
+    try {
+     // State is UNDER_REPLICATED
+      List<ContainerReplicaOp> pendingOpsForContainer =
+          pendingOps.getPendingOps(id);
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOpsForContainer) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
```

Comment:
if an EC container is under-replicated, it may
1 miss some indexes of replica
2 all the indexes of replica is available, but they does not satisfy the placement policy

we should take the seconde scenario into account later

Reply:
We should subtract any indexsPendingAdd from the missingIndexes, as a recovery for some of the indexes may be inflight.

Reply:
We are already passing the flag includePendingAdd as true. Which will remove them right. 
```
  // Now we have a list of missing. Remove any pending add as they should
    // eventually recover.
    if (includePendingAdd) {
      for (Integer i : pendingAdd) {
        missing.remove(i);
      }
    }
```

@JacksonYao287 We will consider the placement policy after the basic reconstruction flows done.